### PR TITLE
support bluegreen release: support workload of deployment and cloneSet

### PR DESCRIPTION
**Pull request by myname4423**
### Ⅰ. Describe what this PR does
1. [mainly] add control logic  for both deployment and cloneSet under `controller/batchrelease`, to support bluegreen style release
2. add e2e for both deployment and cloneSet for bluegreen

### CODE REVIEW

File: pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	patchData.UpdateMinReadySeconds(0)
+	patchData.UpdateMaxSurge(setting.MaxSurge)
+	patchData.UpdateMaxUnavailable(setting.MaxUnavailable)
+	patchData.DeleteAnnotation(v1beta1.OriginalDeploymentStrategyAnnotation)
+	hpa.RestoreHPA(rc.client, rc.object)
+	d := rc.object.DeepCopy()
+	patchData.DeleteAnnotation(util.BatchReleaseControlAnnotation)
+	return rc.client.Patch(context.TODO(), d, patchData)
+}
+func (rc *realController) finalized() bool {
```
Comment (furykerry) [inline pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go:170]: add an empty line

File: api/v1beta1/deployment_types.go [hunk 2]
```diff
@@ -68,10 +68,11 @@ type DeploymentStrategy struct {
 // Similar to DeploymentStrategy, it is an annotation used in workload
 // However, unlike DeploymentStrategy, it is only used to store and restore the user's strategy
 type OriginalDeploymentStrategy struct {
-	// The deployment strategy to use to replace existing pods with new ones.
 	// +optional
-	// +patchStrategy=retainKeys
-	Strategy *apps.DeploymentStrategy `json:"strategy,omitempty" patchStrategy:"retainKeys" protobuf:"bytes,4,opt,name=strategy"`
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,1,opt,name=maxUnavailable"`
```
Comment (furykerry) [inline api/v1beta1/deployment_types.go:72]: OriginalDeploymentStrategy is not part of API, just an internal struct, plz move it out of the api package(maybe to pkg/control?)

File: pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+		release, release.GetObjectKind().GroupVersionKind())))
+	// we use partition = 100% to function as "paused" instead of setting pasued field as true
+	// it is manily to keep consistency with partition style (partition is already set as 100% in webhook)
+	patchData.UpdatePaused(false)
+	maxSurge := intstr.FromInt(1) // select the minimum positive number as initial value
+	maxUnavailable := intstr.FromInt(0)
+	patchData.UpdateMaxSurge(&maxSurge)
+	patchData.UpdateMaxUnavailable(&maxUnavailable)
+	patchData.UpdateMinReadySeconds(v1beta1.MaxReadySeconds)
+	hpa.DisableHPA(rc.client, rc.object)
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go]: done

File: pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+func (rc *realController) Finalize(release *v1beta1.BatchRelease) error {
+	if rc.finalized() {
+		return nil // No need to finalize again.
+	}
+	patchData := patch.NewDeploymentPatch()
+	// continuous release won't run this if-block
+	if release.Spec.ReleasePlan.BatchPartition == nil {
+		setting := util.GetOriginalSetting(rc.object)
+		patchData.UpdatePaused(false)
+		patchData.UpdateMinReadySeconds(0)
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go:178]: done

File: pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+// Add OriginalDeploymentStrategyAnnotation to workload
+func (rc *realController) Initialize(release *v1beta1.BatchRelease) error {
+	if rc.object == nil || control.IsControlledByBatchRelease(release, rc.object) {
+		return nil
+	}
+	setting := util.GetOriginalSetting(rc.object)
+	util.InitOriginalSetting(&setting, rc.object)
+	klog.InfoS("Initialize deployment", "deployment", klog.KObj(rc.object), "setting", util.DumpJSON(&setting))
+
+	d := rc.object.DeepCopy()
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go]: done

File: pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+	}
+	workloadInfo.Status.StableRevision = d.Labels[v1alpha1.DeploymentStableRevisionLabel]
+	return workloadInfo
+}
+
+func (rc *realController) getUpdatedReadyReplicas(d *apps.Deployment) (int32, error) {
+	rss := &apps.ReplicaSetList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(d.Namespace),
+		client.MatchingLabels(d.Spec.Selector.MatchLabels),
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go:257]: done

File: pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+	patchData.UpdateStrategy(apps.DeploymentStrategy{
+		Type: apps.RollingUpdateDeploymentStrategyType,
+		RollingUpdate: &apps.RollingUpdateDeployment{
+			MaxSurge:       &intstr.IntOrString{Type: intstr.Int, IntVal: 1},
+			MaxUnavailable: &intstr.IntOrString{Type: intstr.Int, IntVal: 0},
+		},
+	})
+	patchData.UpdateMinReadySeconds(v1beta1.MaxReadySeconds)
+	patchData.UpdateProgressDeadlineSeconds(utilpointer.Int32(v1beta1.MaxProgressSeconds))
+	hpa.DisableHPA(rc.client, rc.object)
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go]: done

File: pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+	return rc.client.Patch(context.TODO(), d, patchData)
+}
+
+func (rc *realController) UpgradeBatch(ctx *batchcontext.BatchContext) error {
+	if !control.IsReadyForBlueGreenRelease(rc.object) {
+		klog.Warningf("Cannot upgrade batch, because deployment %v doesn't stasify conditions", klog.KObj(rc.object))
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go]: sure

File: pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+		return nil
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go]: ok, Message for batchRelease.status is added now

File: pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go (hunk 1 above)
Comment (myname4423) [reply pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go:190]: sure

File: api/v1beta1/deployment_types.go (hunk 2 above)
Comment (myname4423) [reply api/v1beta1/deployment_types.go]: done

File: pkg/webhook/util/revision/revision.go [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+	}
+	// if equal, it means rollback
+	if bytes.Equal(oldSepc, newSepc) {
+		klog.InfoS("webook detected rollback: the old Spec and new Spec are equal", "oldSpec", oldSepc, "newSpec", newSepc)
+		return false
+	}
+	klog.InfoS("webook detected continuous release: the old Spec and new Spec are not equal", "oldSpec", oldSepc, "newSpec", newSepc)
+	return true
+}
+func GetStablePodTemplate(cli client.Client, cs *kruiseappsv1alpha1.CloneSet) ([]byte, error) {
```
Comment (furykerry) [inline pkg/webhook/util/revision/revision.go:54]: plz add an empty line above

File: pkg/webhook/workload/mutating/workload_update_handler.go [hunk 11]
```diff
@@ -287,7 +291,44 @@ func (h *WorkloadHandler) handleDeployment(newObj, oldObj *apps.Deployment) (boo
...
+					if !util.EqualIgnoreHash(&newObj.Spec.Template, &rs.Spec.Template) {
+						err = fmt.Errorf("successive release is not supported for bluegreen for now, rollback first")
+						klog.Warningf(err.Error())
+						// newObj.Spec.Template = oldObj.Spec.Template
+						return false, err
+					}
+					// for rollback, we set the stable rs minReadySeconds to infinity,
+					// to set pods of the stable rs unavailable, we can prevent the new version pods
+					// from being deleted immediately (we want to keep them until traffic is switched to the stable version)
+					body := fmt.Sprintf(`{"spec":{"minReadySeconds":%v}}`, appsv1beta1.MaxReadySeconds)
```
Comment (furykerry) [inline pkg/webhook/workload/mutating/workload_update_handler.go:311]: why. patch replicaset here? can we just modify the deployment minreadyseconds?

File: pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+	}
+
+	c := util.GetEmptyObjectWithKey(rc.object)
+	setting := control.GetOriginalSetting(rc.object)
+	patchData := patch.NewClonesetPatch()
+	// why we need a simple MinReadySeconds-based status machine? (ie. the if-else block)
+	// It's possible for Finalize to be called multiple times, if error returned is not nil.
+	// if we do all needed operations in a single code block, like, A->B->C, when C need retry,
+	// both A and B will be executed as well, however, operations like restoreHPA cost a lot(which calls LIST API)
+	if rc.object.Spec.MinReadySeconds == v1beta1.MaxReadySeconds {
```
Comment (furykerry) [inline pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go:155]: can we compare the spec.minReadySeconds with the one saved in original setting, if they're not match do the patch. Compare minReadySeconds with an arbitrary constant is not a good idea

File: pkg/controller/batchrelease/control/util.go [hunk 13]
```diff
@@ -112,3 +150,101 @@ func IsCurrentMoreThanOrEqualToDesired(current, desired intstr.IntOrString) bool
 	desiredNum, _ := intstr.GetScaledValueFromIntOrPercent(&desired, 10000000, true)
 	return currentNum >= desiredNum
 }
+
+// GetDeploymentStrategy decode the strategy object for advanced deployment
+// from the annotation "rollouts.kruise.io/original-deployment-strategy"
+func GetOriginalSetting(object client.Object) OriginalDeploymentStrategy {
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/controller/batchrelease/control/util.go]: done

File: pkg/webhook/util/revision/revision.go [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+	return nil, fmt.Errorf("annotation %s is empty", appsv1alpha1.DeploymentStableRevisionLabel)
+}
+
+// get the replicaSet by pod-template-hash label
+func getReplicaSetByLabel(cli client.Client, obj *apps.Deployment, rsLable *map[string]string) (*apps.ReplicaSet, error) {
+	rss := &apps.ReplicaSetList{}
+	listOpts := []client.ListOption{
+		client.InNamespace(obj.Namespace),
+		client.MatchingLabels(obj.Spec.Selector.MatchLabels),
+		client.MatchingLabels(*rsLable),
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/webhook/util/revision/revision.go]: done

File: pkg/webhook/workload/mutating/workload_update_handler.go [hunk 15]
```diff
@@ -369,6 +386,22 @@ func (h *WorkloadHandler) handleCloneSet(newObj, oldObj *kruiseappsv1alpha1.Clon
...
+		in the progress of rollout.
+		Deployment actually has the same problem, however it is possible to bypass this issue for Deployment by setting
+		minReadySeconds for replicaset separately; unfortunately this workaround seems not work for cloneset
+
+		// if rollout.Spec.Strategy.IsBlueGreenRelease() && revision.IsContinuousRelease(h.Client, oldObj, newObj) {
+		// 	err = fmt.Errorf("successive release is not supported for bluegreen for now, rollback first")
+		// 	return false, fmt.Errorf(err.Error())
+		// }
+	*/
+
```
Comment (furykerry) [inline pkg/webhook/workload/mutating/workload_update_handler.go:404]: plz move the comment to the controller logic

File: pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+
+	c := util.GetEmptyObjectWithKey(rc.object)
+	setting, err := control.GetOriginalSetting(rc.object)
+	if err != nil {
+		return errors.NewFatalError(fmt.Errorf("cannot get original setting for cloneset %v: %s from annotation", klog.KObj(rc.object), err.Error()))
+	}
+	patchData := patch.NewClonesetPatch()
+	if rc.object.Spec.MinReadySeconds != setting.MinReadySeconds {
+		// restore the hpa
+		if err := hpa.RestoreHPA(rc.client, rc.object); err != nil {
```
Comment (furykerry) [inline pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go:159]: it is better to patch the workload first before restore HPA.

File: pkg/controller/batchrelease/control/util.go [hunk 17]
```diff
@@ -63,6 +65,42 @@ func IsControlledByBatchRelease(release *v1beta1.BatchRelease, object client.Obj
...
+		if len(o.Spec.UpdateStrategy.Type) > 0 && o.Spec.UpdateStrategy.Type != appsv1alpha1.RecreateCloneSetUpdateStrategyType {
+			return fmt.Errorf("cloneSet strategy type is not ReCreate")
+		}
+		// MinReadySeconds and ProgressDeadlineSeconds must be set
+		if o.Spec.MinReadySeconds != v1beta1.MaxReadySeconds {
+			return fmt.Errorf("cloneSet strategy minReadySeconds is not MaxReadySeconds")
+		}
+
+	default:
+		panic("unsupported workload type to ValidateReadyForBlueGreenRelease function")
```
Comment (furykerry) [inline pkg/controller/batchrelease/control/util.go:99]: can we validate the workload type and rollout version in validating webhook of rollout

File: pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+	if rc.pods != nil {
+		return rc.pods, nil
+	}
+	var err error
+	rc.pods, err = util.ListOwnedPods(rc.client, rc.object)
+	return rc.pods, err
+}
+
+// Initialize prepares the Deployment for the BatchRelease process
+func (rc *realController) Initialize(release *v1beta1.BatchRelease) error {
```
Comment (furykerry) [inline pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go:88]: plz add ut for Initialize, UpgradeBatch and Finalize, which is the most important logic in this patch

File: pkg/util/errors/types.go [hunk 19]
```diff
@@ -0,0 +1,70 @@
+package errors
+
+import (
+	"errors"
+	"fmt"
+)
+
+// BenignError represents a benign error that can be handled or ignored by the caller.
+// It encapsulates information that is non-critical and does not require immediate attention.
+type BenignError struct {
```
Comment (furykerry) [inline pkg/util/errors/types.go:10]: BenignError -> RetryError

Benign is unclear in meaning

File: pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go (hunk 18 above)
Comment (myname4423) [reply pkg/controller/batchrelease/control/bluegreenstyle/deployment/control.go:88]: done

File: pkg/controller/batchrelease/control/util.go (hunk 17 above)
Comment (myname4423) [reply pkg/controller/batchrelease/control/util.go:99]: done

Comment (myname4423) [reply pkg/controller/batchrelease/control/util.go:99]: ps. the v1alpha1 Rollout doesn't has bluegreen field from the begining

File: pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go (hunk 16 above)
Comment (myname4423) [reply pkg/controller/batchrelease/control/bluegreenstyle/cloneset/control.go]: done

File: pkg/webhook/workload/mutating/workload_update_handler.go (hunk 15 above)
Comment (myname4423) [reply pkg/webhook/workload/mutating/workload_update_handler.go:322]: done

File: pkg/util/errors/types.go [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+	return errors.As(err, &benignErr)
+}
+
+func AsBenign(err error, target **BenignError) bool {
+	return errors.As(err, target)
+}
+
+// FatalError represents a fatal error that requires special handling.
+// Such errors are critical and may necessitate logging, alerts, or even program termination.
+type FatalError struct {
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/util/errors/types.go]: done

File: pkg/controller/rollout/rollout_progressing.go [hunk 21]
```diff
@@ -230,6 +236,14 @@ func (r *RolloutReconciler) handleContinuousRelease(c *RolloutContext) error {
 	klog.Infof("rollout(%s/%s) workload continuous publishing canaryRevision from(%s) -> to(%s), then restart publishing",
 		c.Rollout.Namespace, c.Rollout.Name, c.NewStatus.GetCanaryRevision(), c.Workload.CanaryRevision)
 
+	// do nothing for blue-green release
+	if c.Rollout.Spec.Strategy.IsBlueGreenRelease() {
+		cond := util.GetRolloutCondition(*c.NewStatus, v1beta1.RolloutConditionProgressing)
+		cond.Message = "[warning] new version released in progress of blue-green release, please rollback first"
+		c.NewStatus.Message = cond.Message
+		return utilerrors.NewFatalError(fmt.Errorf("cannot do continuous release for blue-green release, rollback firstly"))
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/controller/rollout/rollout_progressing.go]: done

File: pkg/controller/rollout/rollout_progressing.go [hunk 22]
```diff
@@ -110,7 +111,12 @@ func (r *RolloutReconciler) reconcileRolloutProgressing(rollout *v1beta1.Rollout
 	case v1alpha1.ProgressingReasonInRolling:
 		klog.Infof("rollout(%s/%s) is Progressing, and in reason(%s)", rollout.Namespace, rollout.Name, cond.Reason)
 		err = r.doProgressingInRolling(rolloutContext)
-		if err != nil {
+		if utilerrors.IsFatal(err) {
+			// For fatal errors, do not retry as it wastes resources and has no effect.
+			// therefore, we don't propagate the error, but just log it.
+			// user should do sth instead, eg. for bluegreen continuous release scenario, user should do rollback
+			klog.Warningf("rollout(%s/%s) doProgressingInRolling error(%s)", rollout.Namespace, rollout.Name, err.Error())
```
[the remark this answers is not in the document]
Comment (myname4423) [reply pkg/controller/rollout/rollout_progressing.go:118]: done

**Pull request merged**